controllers/statemachine/machine.go

```diff
@@ -0,0 +1,277 @@
+/*
+Copyright 2022 The VolSync authors.
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU Affero General Public License as published
+by the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Affero General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License
+along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+package statemachine
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-logr/logr"
+	cron "github.com/robfig/cron/v3"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+// replicationState is the different states that replication object can be in
+type replicationState string
+
+const (
+	initialState       replicationState = "Initial"
+	synchronizingState replicationState = "Synchronizing"
+	cleaningUpState    replicationState = "CleaningUp"
+)
+
+// triggerType represents the different ways we can trigger data synchronization
+type triggerType string
+
+const (
+	scheduleTrigger triggerType = "ScheduleTrigger"
+	manualTrigger   triggerType = "ManualTrigger"
+	noTrigger       triggerType = "NoTrigger"
+)
+
+// Run the state machine to reconcile the ReplicationController
+func Run(ctx context.Context, r ReplicationMachine, l logr.Logger) (ctrl.Result, error) {
+	// Set out-of-sync metrics flag if necessary
+	if r.LastSyncTime() == nil {
+		r.SetOutOfSync(true)
+	} else {
+		missed, err := missedDeadline(r)
+		if err != nil {
+			setConditionError(r, l, err)
+			return ctrl.Result{}, err
+		}
+		if missed {
+			r.SetOutOfSync(true)
+		}
+	}
+
+	var result ctrl.Result
+	var err error
+	switch currentState(r) {
```

Comment:
I might be missing something here - but do we have a potential to start synchronizing out of schedule here?

I think previously there was always a check to see if we should sync (similar to the new `shouldSync()` func) prior to synchronizing, but I don't see it anymore.

If the reconcile is triggered by a change to some resource the RD/RS owns, or even something like the controller restarting (and I think there was a regular sync interval added at some point too just in case), then it seems we could enter into synchronizing (by running `doSynchronizingState()`) even when technically we're in "idle" state (for example when the next sync time hasn't hit yet).  Edit:  I suppose this could happen with trigger as well - if the trigger hasn't changed, we could possibly run a second sync with the same trigger?

Reply:
We (are supposed to) stay locked in a state until one of the transitionToX() functions gets called.
"Idle" is really just cleanup cleanup. We stay in cleanup, but report that we're idle (L130-L144).
So, once we're in cleanup, if the mover says it's complete, then we call shouldSync(), and that's what moves us to synchronizing (L125). Extra reconciles just cause additional checks on the resources and return Complete but have no effect unless we shouldSync().

Reply:
aha, yes that's the part I didn't catch - after completing synchronizing the lastSyncStartTime should get cleared, so it will go to cleanup instead of synchronizing again.  Thx for the explanation